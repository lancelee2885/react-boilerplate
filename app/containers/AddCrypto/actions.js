/*
 *
 * AddCrypto actions
 *
 */

import {
  CHANGE_FORM_DATA,
  SUBMIT_FORM_DATA,
  SUBMIT_FORM_DATA_ERROR,
  SUBMIT_FORM_DATA_SUCCESS,
} from './constants';

/**
 * Changes the input fields of the form
 */
export function changeFormData(formData) {
  return {
    type: CHANGE_FORM_DATA,
    formData,
  };
}

/**
 * Changes the input fields of the form
 */
export function submitFormData() {
  return {
    type: SUBMIT_FORM_DATA,
  };
}

/**
 * Changes the input fields of the form
 */
export function formDataSubmitted(formData) {
  return {
    type: SUBMIT_FORM_DATA_SUCCESS,
    formData,
  };
}

/**
 * Changes the input fields of the form
 */
export function formDataError(err) {
  return {
    type: SUBMIT_FORM_DATA_ERROR,
    err,
  };
}
