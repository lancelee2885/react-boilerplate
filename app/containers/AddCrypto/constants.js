/*
 *
 * AddCrypto constants
 *
 */

export const CHANGE_FORM_DATA = 'app/AddCrypto/CHANGE_FORM_DATA';
export const SUBMIT_FORM_DATA = 'app/AddCrypto/SUBMIT_FORM_DATA';
export const SUBMIT_FORM_DATA_SUCCESS =
  'app/AddCrypto/SUBMIT_FORM_DATA_SUCCESS';
