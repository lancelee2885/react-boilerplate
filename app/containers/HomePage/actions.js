/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_CRYPTO,
  LOAD_CRYPTO_ERROR,
  LOAD_CRYPTO_SUCCESS,
} from './constants';

/**
 * Load the crypto from backend, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_CRYPTO
 */
export function loadCryptos() {
  return {
    type: LOAD_CRYPTO,
  };
}

/**
 * Dispatched when the cryptocurrencies are loaded by the request saga
 *
 * @param  {array} cryptos The crypto data
 *
 * @return {object}      An action object with a type of LOAD_CRYPTO_SUCCESS passing the cryptos
 */
export function cryptosLoaded(cryptos) {
  return {
    type: LOAD_CRYPTO_SUCCESS,
    cryptos,
  };
}

/**
 * Dispatched when loading the cryptocurrencies fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_CRYPTO_ERROR passing the error
 */
export function cryptossLoadingError(error) {
  return {
    type: LOAD_CRYPTO_ERROR,
    error,
  };
}
