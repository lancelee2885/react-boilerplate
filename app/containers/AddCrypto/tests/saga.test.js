/**
 * Tests for AddCrypto sagas
 */

import { call, put, takeLatest } from 'redux-saga/effects';

import { SUBMIT_FORM_DATA } from '../constants';
import { formDataSubmitted, formDataError } from '../actions';

import loadNewData, { createNewCrypto, submitToServer } from '../saga';

/* eslint-disable redux-saga/yield-effects */
describe('createNewCrypto Saga', () => {
  let createCryptosGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    createCryptosGenerator = createNewCrypto();

    const callDescriptor = createCryptosGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the submitToServer action if it sends the data successfully', () => {
    const callDescriptor = createCryptosGenerator.next('test');
    expect(callDescriptor.value).toEqual(call(submitToServer, 'test'));
  });

  it('should dispatch the formDataSubmitted action if it sends the data successfully', () => {
    createCryptosGenerator.next('test');
    const putDescriptor = createCryptosGenerator.next('test');
    expect(putDescriptor.value).toEqual(put(formDataSubmitted('test')));
  });

  it('should call the formDataError action if the response errors', () => {
    const response = new Error('Some error');
    createCryptosGenerator.next();
    const putDescriptor = createCryptosGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(formDataError(response)));
  });
});

describe('loadNewData Saga', () => {
  const newDataSaga = loadNewData();

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = newDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(SUBMIT_FORM_DATA, createNewCrypto),
    );
  });
});
