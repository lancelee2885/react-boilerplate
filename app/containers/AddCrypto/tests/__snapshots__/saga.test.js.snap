// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNewCrypto Saga should call the formDataError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createNewCrypto Saga should dispatch the formDataSubmitted action if it sends the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createNewCrypto Saga should dispatch the submitToServer action if it sends the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
